interpreter/dispatcher.go

```diff
@@ -0,0 +1,207 @@
+		functions: make(map[string]map[int][]*overload)}
+}
+
+// Helper types for tracking overloads by various dimensions.
+type overloadMap map[string]*overload
+type overloadMapByFunctionAndArgCount map[string]map[int][]*overload
+
+type defaultDispatcher struct {
+	overloads overloadMap
+	functions overloadMapByFunctionAndArgCount
+}
+
+var _ Dispatcher = &defaultDispatcher{}
+
+func (d *defaultDispatcher) Add(overloads ...*functions.Overload) error {
+	for _, o := range overloads {
+		// Determine the arg count and type from the overload signature.
+		iface := reflect.TypeOf(o.Signature)
+		argCount := iface.NumIn()
+		argTypes := make([]types.Type, argCount)
+		for i := 0; i < argCount; i++ {
+			refType := iface.In(i)
+			var argType types.Type = nil
+			// If the arg type is an interface{}, this is equivalent to dyn.
+			if refType.Kind() == reflect.Interface {
+				argType = types.DynType
+				// Otherwise, instantiate a zero-value of the argument and inspect
+				// the value type.
+			} else {
+				refVal := reflect.New(refType).Elem()
+				if argTypeVal, found := types.TypeOf(refVal.Interface()); found {
+					argType = argTypeVal
+				}
+			}
+			// Set the arg type for the argument at the ordinal 'i' to the
+			// type that was resolved, otherwise, return an error.
+			if argType != nil {
+				argTypes[i] = argType
+			} else {
+				return fmt.Errorf("unrecognized type '%T'"+
+					" in function signature for overload '%s'",
+					refType, o.Name)
+			}
+		}
+		overloadRef := &overload{o.Function,
+			o.Name,
+			argCount,
+			argTypes,
+			o.Impl}
+		// Add the overload unless an overload of the same name has already
+		// been provided before.
+		if _, found := d.overloads[o.Name]; found {
+			return fmt.Errorf("overload already exists '%s'", o.Name)
+		}
+		// Index the overload by function and by arg count.
+		// TODO: consider indexing by abstract type.
+		d.overloads[o.Name] = overloadRef
+		if byFunction, found := d.functions[o.Function]; !found {
+			byFunction = make(map[int][]*overload)
+			byFunction[argCount] = []*overload{overloadRef}
+			d.functions[o.Function] = byFunction
+		} else if byArgCount, found := byFunction[argCount]; !found {
+			byFunction[argCount] = []*overload{overloadRef}
+		} else {
+			byFunction[argCount] = append(byArgCount, overloadRef)
+		}
+	}
+	return nil
+}
+
+func (d *defaultDispatcher) Dispatch(ctx *CallContext) (interface{}, error) {
+	if overload, err := d.findOverload(ctx); err == nil {
+		return overload.impl(ctx.args...)
+	} else {
+		return nil, err
+	}
+}
+
+// Find the overload that matches the call context. There should be only one,
+// otherwise the call is ambiguous and a runtime error produced.
+func (d *defaultDispatcher) findOverload(ctx *CallContext) (*overload, error) {
+	// TODO: Add location metadata to overloads.
+	if overloadId, found := ctx.Overload(); found {
+		if match, found := d.overloads[overloadId]; found {
+			return match, nil
+		} else {
```

Comment:
Ditto. Suggest dropping `else` statements after `return`.  Similarly, many more below in this PR.

Reply:
Done